File: mysql-code-generator/src/main/java/org/xi/quick/codegenerator/command/GeneratorCommand.java
package org.xi.quick.codegenerator.command;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.CommandLineRunner;
import org.springframework.core.annotation.Order;
import org.springframework.stereotype.Component;
import org.xi.quick.codegenerator.functionalinterface.BinaryConsumer;
import org.xi.quick.codegenerator.service.GeneratorService;
import org.xi.quick.codegenerator.service.TableService;

import java.util.*;
import java.util.concurrent.atomic.AtomicReference;
import java.util.regex.Pattern;

/**
 * @author 郗世豪（rarexixi@outlook.com）
 * @date 2018/01/01 10:53
 */
@Component
@Order(100)
public class GeneratorCommand implements CommandLineRunner {

    Logger logger = LoggerFactory.getLogger(GeneratorCommand.class);

    @Autowired
    TableService tableService;

    @Autowired
    GeneratorService generatorService;

    @Override
    public void run(String... strings) throws Exception {

        Set<String> tableNameSet = tableService.getAllTableNameList();

        Scanner sc = new Scanner(System.in);

        printUsages();
        while (sc.hasNextLine()) {
            try {
                processLine(sc, tableNameSet);
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                printUsages();
            }
        }
    }

    private static void printUsages() {
        System.out.println("操作说明:");
        System.out.println("\tga/da:                        生成/删除所有文件(gen/del all)");
        System.out.println("\tgb/db:                        生成/删除基本类型的相关文件(gen/del base)");
        System.out.println("\tgag/dag:                      生成聚合相关文件(gen/del aggregate)");
        System.out.println("\tgat/dat:                      生成/删除所有表的相关文件(gen/del all table)");
        System.out.println("\tgen/del table [table2...]:    根据表名生成/删除相关文件");
        System.out.println("\ts (show):                     显示所有表名");
        System.out.println("\tq (quit):                     退出");
        System.out.print("请输入命令 : ");
    }

    private void processLine(Scanner sc, Set<String> tableNameSet) {
        String cmd = sc.next().toLowerCase();
        String[] args;
        switch (cmd) {
            case "ga":
                generatorService.generateAll();
                generatorService.generateAllOnce();
                generatorService.generateAllAggregate();
                return;
            case "da":
                generatorService.deleteAll();
                generatorService.deleteAllOnce();
                generatorService.deleteAllAggregate();
                return;
            case "gb":
                generatorService.generateAllOnce();
                return;
            case "db":
                generatorService.deleteAllOnce();
                return;
            case "gag":
                generatorService.generateAllAggregate();
                return;
            case "dag":
                generatorService.deleteAllAggregate();
                return;
            case "gat":
                generatorService.generateAll();
                return;
            case "dat":
                generatorService.deleteAll();
                return;
            case "gen":
                generate(tableNameSet, getArgs(sc));
                return;
            case "del":
                delete(tableNameSet, getArgs(sc));
                return;
            case "show":
            case "s":
                showTables(tableNameSet);
                return;
            case "quit":
            case "q":
                System.exit(0);
                return;
            default:
                logger.error("[错误] 未知命令:" + cmd);
                return;
        }
    }

    private void generate(Set<String> tableNameSet, String[] tables) {

        invoke(tableNameSet, tables, false, (tablesToInvoke, tablesNotExist) -> {

            generatorService.generateSet(tablesToInvoke);

            if (!tablesNotExist.isEmpty()) {
                logger.warn("表" + String.join(",", tablesNotExist) + "不存在");
            }
        });
    }

    private void delete(Set<String> tableNameSet, String[] tables) {

        invoke(tableNameSet, tables, true, (tablesToInvoke, tablesNotExist) -> generatorService.deleteSet(tablesToInvoke));
    }

    private void invoke(Set<String> tableNameSet, String[] tables, boolean isDelete, BinaryConsumer<Set<String>, Set<String>> consumer) {

        if (tables.length == 0) return;

        Set<String> tablesToInvoke = new HashSet<>();
        Set<String> tablesNotExist = new HashSet<>();

        for (String table : tables) {
            if (tableNameSet.contains(table)) {
                tablesToInvoke.add(table);
            } else if (table.contains("*")) {
                String tableRegex = '^' + table.replace("*", ".*") + '$';
                tableNameSet.forEach(it -> {
                    if (Pattern.matches(tableRegex, it)) {
                        tablesToInvoke.add(it);
                    }
                });
            } else if (isDelete) {
                tablesToInvoke.add(table);
            }  else {
                tablesNotExist.add(table);
                continue;
            }
        }

        consumer.accept(tablesToInvoke, tablesNotExist);
    }

    private void showTables(Set<String> tableNameSet) {
        int maxLength = 0;
        for (String s : tableNameSet) {
            if (s.length() > maxLength) maxLength = s.length();
        }
        String formatString = "%-" + (maxLength + 4) + "s";
        AtomicReference<Integer> index = new AtomicReference<>(0);
        tableNameSet.forEach(it -> {
            System.out.printf(formatString, it);
            index.set(index.get() + 1);
            if (index.get() % 4 == 0) {
                System.out.println();
            }
        });
    }

    String[] getArgs(Scanner sc) {
        String line = sc.nextLine();
        if (line == null) return new String[0];
        StringTokenizer tokenizer = new StringTokenizer(line, " ");
        List result = new ArrayList();

        while (tokenizer.hasMoreElements()) {
            Object s = tokenizer.nextElement();
            result.add(s);
        }
        return (String[]) result.toArray(new String[result.size()]);
    }
}

File: mysql-code-generator-kt/src/main/kotlin/org/xi/quick/codegeneratorkt/command/GeneratorCommand.kt
package org.xi.quick.codegeneratorkt.command

import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.CommandLineRunner
import org.springframework.stereotype.Component
import org.xi.quick.codegeneratorkt.service.GeneratorService
import org.xi.quick.codegeneratorkt.service.TableService
import java.util.*

@Component
class GeneratorCommand : CommandLineRunner {

    internal var logger = LoggerFactory.getLogger(GeneratorCommand::class.java)

    @Autowired
    lateinit var tableService: TableService

    @Autowired
    lateinit var generatorService: GeneratorService

    @Throws(Exception::class)
    override fun run(vararg strings: String) {

        val tableNameSet = tableService.getAllTableNameList()

        val sc = Scanner(System.`in`)

        printUsages()
        while (sc.hasNextLine()) {
            try {
                processLine(sc, tableNameSet)
            } catch (e: Exception) {
                e.printStackTrace()
            } finally {
                printUsages()
            }
        }
    }

    private fun printUsages() {
        var introduction = """
操作说明:
    ga/da:                        生成/删除所有文件
    gb/db:                        生成/删除基本类型的相关文件
    gag/dag:                      生成聚合相关文件
    gat/dat:                      生成/删除所有表的相关文件
    gen/del table [table2...]:    根据表名生成/删除相关文件(多个表之间空格分开，支持*通配符)
    s (show):                     显示所有表名
    q (quit):                     退出
请输入命令: """
        print(introduction)
    }

    private fun processLine(sc: Scanner, tableNameSet: Set<String>) {
        val cmd = sc.next().toLowerCase()
        val args = getArgs(sc)
        when (cmd) {
            "ga" -> {
                generatorService.gen()
                generatorService.genBase()
                generatorService.genAggr()
            }
            "da" -> {
                generatorService.del()
                generatorService.delBase()
                generatorService.delAggr()
            }
            "gb" -> generatorService.genBase()
            "db" -> generatorService.delBase()
            "gag" -> generatorService.genAggr(*args)
            "dag" -> generatorService.delAggr()
            "gat" -> generatorService.gen()
            "dat" -> generatorService.del()
            "gen" -> gen(tableNameSet, args)
            "del" -> del(tableNameSet, args)
            "show", "s" -> showTables(tableNameSet)
            "quit", "q" -> System.exit(0)
            else -> logger.error("[错误] 未知命令:$cmd")
        }
    }

    private fun gen(tableNameSet: Set<String>, tables: Array<String>) {

        if (tables.isEmpty()) return

        invoke(tableNameSet, tables, false) { tablesToInvoke, tablesNotExist ->
            apply {
                if (tablesToInvoke.isNotEmpty())
                    generatorService.gen(*tablesToInvoke)
                if (tablesNotExist.isNotEmpty()) {
                    logger.warn("表" + tablesNotExist.joinToString(",") + "不存在或者没有配置")
                }
            }
        }
    }

    private fun del(tableNameSet: Set<String>, tables: Array<String>) {

        if (tables.isEmpty()) return

        invoke(tableNameSet, tables, true) { tablesToInvoke, tablesNotExist ->
            apply {
                generatorService.del(*tablesToInvoke)
            }
        }
    }

    private fun invoke(tableNameSet: Set<String>, tables: Array<String>, isDelete: Boolean, consume: (Array<String>, Array<String>) -> Unit) {

        var tablesToInvoke = HashSet<String>()
        val tablesNotExist = HashSet<String>()

        for (table in tables) {
            if (tableNameSet.contains(table)) {
                tablesToInvoke.add(table)
            } else if (table.contains("*")) {
                var tableRegex = ('^' + table.replace("*", ".*") + '$').toRegex()
                tableNameSet.forEach {
                    if (tableRegex.matches(it)) {
                        tablesToInvoke.add(it)
                    }
                }
            } else if (isDelete) {
                tablesToInvoke.add(table)
            } else {
                tablesNotExist.add(table)
                continue
            }
        }

        consume(tablesToInvoke.toTypedArray(), tablesNotExist.toTypedArray())
    }

    private fun showTables(tableNameSet: Set<String>) {
        var maxLength = tableNameSet.map { it.length }.max()?.plus(4)
        var formatString = "%-" + maxLength + "s"
        var index = 0
        tableNameSet.forEach {
            print(String.format(formatString, it))
            if (++index % 4 == 0) {
                println()
            }
        }
    }

    internal fun getArgs(sc: Scanner): Array<String> {
        val line = sc.nextLine() ?: return arrayOf()
        val tokenizer = StringTokenizer(line, " ")
        val result = ArrayList<String>()

        while (tokenizer.hasMoreElements()) {
            val s = tokenizer.nextElement()
            result.add(s.toString())
        }
        return result.toTypedArray()
    }
}
